sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,428 @@
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class _VcfSource(filebasedsource.FileBasedSource):
+  """A source for reading VCF files.
+
+  Parses VCF files (version 4) using PyVCF library. If file_pattern specifies
+  multiple files, then the header from each file is used separately to parse
+  the content. However, the output will be a uniform PCollection of
+  :class:`Variant` objects.
+  """
+
+  DEFAULT_VCF_READ_BUFFER_SIZE = 65536  # 64kB
+
+  def __init__(self,
+               file_pattern,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfSource, self).__init__(file_pattern,
+                                     compression_type=compression_type,
+                                     validate=validate)
+
+    self._header_lines_per_file = {}
+    self._file_pattern = file_pattern
+    self._compression_type = compression_type
+    self._buffer_size = buffer_size
+    self._validate = validate
+
+  def read_records(self, file_name, range_tracker):
+    record_iterator = _VcfSource._VcfRecordIterator(
+        file_name,
+        range_tracker,
+        self._file_pattern,
+        0,  # min_bundle_size
+        self._compression_type,
+        True,  # strip_trailing_newlines
+        coders.StrUtf8Coder(),  # coder
+        buffer_size=self._buffer_size,
+        validate=self._validate,
+        skip_header_lines=0)
+
+    # Convert iterator to generator to abstract behavior
+    for line in record_iterator:
+      yield line
+
+  class _VcfRecordIterator(object):
+    """An Iterator for processing a single VCF file."""
+
+    def __init__(self,
+                 file_name,
+                 range_tracker,
+                 file_pattern,
+                 min_bundle_size,
+                 compression_type,
+                 strip_trailing_newlines,
+                 coder,
+                 **kwargs):
+      self._header_lines = []
+      self._last_record = None
+      self._file_name = file_name
+
+      text_source = TextSource(
+          file_pattern,
+          min_bundle_size,
+          compression_type,
+          strip_trailing_newlines,
+          coder,
+          header_processor_fns=(lambda x: x.startswith('#'),
+                                self._store_header_lines),
+          **kwargs)
+
+      self._text_lines = text_source.read_records(self._file_name,
+                                                  range_tracker)
+      try:
+        self._vcf_reader = vcf.Reader(fsock=self._create_generator())
+      except SyntaxError as e:
+        raise ValueError('Invalid VCF header %s' % str(e))
+
+    def _store_header_lines(self, header_lines):
+      self._header_lines = header_lines
+
+    def _create_generator(self):
+      header_processed = False
+      for record in self._text_lines:
```

Comment:
<!--new_thread; commit:63511ddb8caf9e5c12888ef8a4d1071e0ea6c168; resolved:0-->
s/record/text_line. I think it's clearer if we use term "text_line" for lines returned by text files and record for Variant records generated by this source.

Reply:
Done.